### ISSUE TITLE
【社区任务】融合线性交叉熵损失算子 Ascend 适配

### ISSUE THREAD

**Issue by stwilliam503388-creator** [triton-lang/triton-ascend#1816]
## 认领方式

在当前Issue下留言「认领这个任务」
我们会及时分配并@您
一位开发者只能同时领取一个任务，启动验收后才能领取下一个，同时需要开发者在昇腾社区任务广场的对应任务完成报名
开发者需要定期和对接人沟通进展，除特殊情况外，如果连续两个星期没有进展，任务将被释放

## 社区任务广场

[【Triton-Ascend】融合线性交叉熵损失算子 Ascend 适配](https://www.hiascend.com/activities/task-center/details/e7b3bf2e2478474b83405d84ad4fbe5f)

---

## 任务类型

算子开发

## 任务描述

在昇腾NPU上基于triton-ascend编程语言，完成上游 [fla-org/flash-linear-attention](https://github.com/fla-org/flash-linear-attention) 中融合交叉熵及融合线性交叉熵损失算子的昇腾亲和支持，包括 `fused_cross_entropy`、`fused_linear_cross_entropy` 及其 softcap 变体，面向大词表语言模型训练场景。完成算子设计、开发、测试及上游适配全流程工作，最终提交上游 PR 并被接受。

## 任务标准

1. 在 triton-ascend 3.2.2 + torch_npu 环境下，运行 pytest tests/modules/test_cross_entropy.py 全量用例通过（0 failed、0 error），不允许修改测试用例中的断言与容差；
2. 平台覆盖：在 Ascend A2、A3、950 平台上均可运行通过（各平台均 0 failed、0 error）；如因平台硬件差异存在限制，须在 PR 中说明原因并提供适配或规避方案；
3. 消除 NPU 上 aicore / vector core 超时（错误码 507014 / 507034）及编译器崩溃，kernel 可在规定时间内正常执行退出；
4. 数值精度满足测试内 assert_close 容差要求，前向与反向（含梯度）结果均正确；
5. 固定长度与 varlen、fp16 与 fp32 等全部参数化配置均通过；
6. 不引入回归：同目录下其余已通过用例保持通过；
7. 提交修复 PR，附根因分析、修改说明与本地验证命令及结果。

### 功能实现要求

1. 基于 triton-ascend 3.2.2 完成 `fused_cross_entropy`、`fused_linear_cross_entropy` 及其 softcap 变体的昇腾亲和适配，保证算子功能与上游实现一致。
2. 支持融合交叉熵损失及融合线性交叉熵损失的前向和反向计算，保证输出及梯度结果正确。
3. 支持大词表场景，包括词表大小 `V=32000` 和 `V=100000`。
4. 支持 `reduction='mean'`、不同 `logit_scale` 及 `accumulate_grad_in_fp32` 等测试要求的配置。
5. 支持 logit softcap，包括 `softcap=30.0` 场景。
6. 支持 BF16 等测试要求的数据类型，并针对大词表场景进行昇腾NPU计算及访存优化。
7. Kernel 实现需符合 Triton-Ascend 编程规范，并针对昇腾NPU的计算及访存特性进行适配和优化。
8. 完成适配后向上游 `fla-org/flash-linear-attention` 提交 PR，并确保 PR 被接受。

### 参数说明

| 参数名                     | 输入／输出/属性 | 描述                      | 数据类型      | 数据格式 | 维度（shape） |
| ----------------------- | -------- | ----------------------- | --------- | ---- | --------- |
| logits                  | 输入       | 融合交叉熵损失计算的输入 logits     | BF16      | ND   | [B×T,V]   |
| x                       | 输入       | 融合线性交叉熵损失的输入特征          | BF16      | ND   | [B×T,D]   |
| weight                  | 输入       | 线性层权重，用于计算 logits       | BF16      | ND   | [V,D]     |
| bias                    | 输入       | 线性层偏置                   | BF16      | ND   | [V]       |
| target                  | 输入       | 交叉熵计算的目标类别索引            | INT64     | ND   | [B×T]     |
| logit_scale             | 属性       | 对 logits 进行缩放的系数        | FLOAT     | -    | -         |
| logit_softcapping       | 属性       | 对 logits 进行 softcap 的系数 | FLOAT     | -    | -         |
| reduction               | 属性       | 损失归约方式                  | STRING    | -    | -         |
| accumulate_grad_in_fp32 | 属性       | 是否以 FP32 累积梯度           | BOOL      | -    | -         |
| out                     | 输出       | 交叉熵损失计算结果               | BF16      | ND   | [1]       |
| dx                      | 输出       | 对输入特征 x 的梯度             | BF16/FP32 | ND   | [B×T,D]   |
| dw                      | 输出       | 对权重 weight 的梯度          | BF16/FP32 | ND   | [V,D]     |
| db                      | 输出       | 对偏置 bias 的梯度            | BF16/FP32 | ND   | [V]       |

### 约束说明

1. `fused_cross_entropy` 的 `logits` shape 为 `[B×T,V]`，其中 B 为 Batch Size，T 为 Sequence Length，V 为词表大小。
2. `fused_linear_cross_entropy` 的输入 `x` shape 为 `[B×T,D]`，`weight` shape 为 `[V,D]`，`bias` shape 为 `[V]`。
3. 支持词表大小 `V=32000`、`V=100000` 等大词表场景。
4. `target` 中值为 `-100` 的位置按照交叉熵的 ignore index 语义处理。
5. 支持 `logit_scale`、`logit_softcapping` 以及 `accumulate_grad_in_fp32` 等属性配置。
6. softcap 场景下需与参考实现的 logits softcap 计算保持一致。
7. Forward 与 Backward 计算结果均需与上游参考实现保持一致。
8. 不得针对测试用例中的单一 shape 进行硬编码，需要具备输入 shape 泛化能力。
9. 不得修改测试用例中的断言、精度容差或通过跳过测试规避问题。

## 测试标准

在 triton-ascend 3.2.2 + torch_npu 环境下，基于上游 `tests/modules/test_cross_entropy.py` 测试文件运行：

```bash
pytest tests/modules/test_cross_entropy.py
```

要求全量用例通过，`0 failed、0 error`，不允许修改测试用例中的断言与容差。

## PR 申请合入

验收通过后，在 Triton-Ascend-Kernels 开源仓提交 PR 申请，申请将开发完成的算子合入（https://gitcode.com/Ascend/triton-ascend-kernels/tree/experimental）。

## 参考资料

1. [fla-org/flash-linear-attention](https://github.com/fla-org/flash-linear-attention)

2. [Triton-Ascend 文档](https://triton-ascend.readthedocs.io/zh-cn/latest/quick_start.html)

3. [Triton-Ascend 代码样例](https://gitcode.com/Ascend/triton-ascend/tree/main/docs/zh/examples)

4. [Triton-Ascend-Kernels](https://gitcode.com/Ascend/triton-ascend-kernels)

5. 上游 Cross Entropy 测试文件：`tests/modules/test_cross_entropy.py`

## 环境获取

依据 [fla-org/flash-linear-attention](https://github.com/fla-org/flash-linear-attention) 仓库说明获取开发环境，配置 Triton-Ascend==3.2.2、torch-npu==2.9.0 及 CANN 9.1.0 环境。

**Comment by lowdy1** [comment]
[cross_entropy.md](https://github.com/user-attachments/files/31590331/cross_entropy.md)

**Comment by stwilliam503388-creator** [comment]
此任务已释放/关闭，从主任务池表格中移除。
